ssh2/session.pyx

```diff
@@ -45,18 +50,49 @@ LIBSSH2_HOSTKEY_TYPE_UNKNOWN = c_ssh2.LIBSSH2_HOSTKEY_TYPE_UNKNOWN
 LIBSSH2_HOSTKEY_TYPE_RSA = c_ssh2.LIBSSH2_HOSTKEY_TYPE_RSA
 LIBSSH2_HOSTKEY_TYPE_DSS = c_ssh2.LIBSSH2_HOSTKEY_TYPE_DSS
 
+cdef ssize_t _send_callback(int fd,
+                            char* buffer,
+                            size_t length,
+                            int flags,
+                            void** abstract) with gil:
+    global errno
+    try:
+        bp = PyBytes_FromStringAndSize(buffer, length)
+        res = (<Session>deref(abstract))._send_callback(fd, bp, flags)
+        return length
+    except Exception as e:
+        return -e.errno
+
+cdef ssize_t _recv_callback(int fd,
+                            char* buffer,
+                            size_t length,
+                            int flags,
+                            void** abstract) with gil:
+    global errno
+    try:
+        data = (<Session>deref(abstract))._recv_callback(fd, length, flags)
```

Comment:
The function is globally scoped, so `abstract` is the only way that the function has access to the callback method pointer.

Reply:
I guess I am confused by your question.  Do you mean pass the callback directly into `libssh2_session_callback_set`?

Reply:
No, meant passing the python function directly into the `_send_callback` C function, with python function casted as `<void *>`. It can then get called directly in `_send_callback`, casted back to a python `<object>` instead of via `(<Session>deref(abstract))._send_callback`. See [here for an example](https://github.com/ParallelSSH/ssh-python/blob/master/ssh/callbacks.pyx).

Passing in the session object reference to the C callback function then calling the Session's callback function attribute seems a bit redundant - unless there is a reason for it.

Reply:
I think it just makes things more pythonic; returning strings instead of lengths and strings, plus the `except -1` stuff.